README
NAME
    WWW::FMyLife - Obtain FMyLife.com anecdotes via API

VERSION
    Version 0.14

SYNOPSIS
    THIS MODULE IS STILL UNDER INITIAL DEVELOPMENT! BE WARNED!

        use WWW::FMyLife;

        my $fml = WWW::FMyLife->new();
        print map { "Items: $_\n" } $fml->last( { as => text' } );

DESCRIPTION
    This module fetches FMyLife.com (FML) anecdotes, comments, votes and
    more via API, comfortably and in an extensible manner.

        my @items = $fml->top_daily();
        foreach my $item (@items) {
            my $item_id      = $item->id;
            my $item_content = $item->content;
            print "[$item_id] $item_content\n";
        }

        print $fml->random()->text, "\n";
        ...

EXPORT
    This module exports nothing.

SUBROUTINES/METHODS
  last()
    Fetches the last quotes. Can accept a hashref that indicates the
    formatting:

        # returns an array of WWW::FMyLife::Item objects
        $fml->last();

        # or more explicitly
        $fml->last( { as => 'object' } ); # same as above
        $fml->last( { as => 'text'   } ); # returns an array of text anecdotes
        $fml->last( { as => 'data'   } ); # returns an array of hashes of anecdotes

    You can also specify which page you want:

        # return 1st page
        my @last = fml->last();

        # return 5th page
        my @last = $fml->last(5);

        # same
        my @last = $fml->last( { page => 5 } );

    And options can be mixed:

        my @not_so_last = $fml->last( { as => 'text', page => 50 } );

  random
    This method gets a single random quote as an object.

  top
    This method works the same as the last() method, only it fetches the top
    quotes.

    This method, as for its variations, can format as an object, text or
    data.

  top_day
    This method works the same as the last() method, only it fetches the top
    quotes.

    This specific variant fetches the top anecdotes from the last day.

  top_week
    This method works the same as the last() method, only it fetches the top
    quotes.

    This specific variant fetches the top anecdotes from the last week.

  top_month
    This method works the same as the last() method, only it fetches the top
    quotes.

    This specific variant fetches the top anecdotes from the last month.

  flop
    Fetches the flop quotes.

    This method, as for its variations, can format as an object, text or
    data.

  flop_day
    Fetches the flop quotes of the day.

  flop_week
    Fetches the flop quotes of the week.

  flop_month
    Fetches the flop quotes of the month.

  credentials( $username, $password ) (NOT YET FULLY IMPLEMENTED)
    WARNING: THIS HAS NOT YET BEEN IMPLEMENTED.

    THE TESTS HAVE BEEN DISABLED FOR NOW, PLEASE WAIT FOR A MORE ADVANCED
    VERSION.

    Sets credentials for members.

        $fml->credentials( 'foo', 'bar' );

        # same thing
        $fml->username('foo');
        $fml->password('bar');

AUTHOR
    Sawyer X (XSAWYERX), "<xsawyerx at cpan.org>"

    Tamir Lousky (TLOUSKY), "<tlousky at cpan.org>"

DEPENDENCIES
    Moose

    XML::Simple

    LWP::UserAgent

BUGS AND LIMITATIONS
    Please report any bugs or feature requests to "bug-www-fmylife at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-FMyLife>.

    You can also use the Issues Tracker on Github @
    <http://github.com/xsawyerx/www-fmylife/issues>.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc WWW::FMyLife

    You can also look for information at:

    *   Our Github!

        <http://github.com/xsawyerx/www-fmylife/tree/master>

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-FMyLife>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/WWW-FMyLife>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/WWW-FMyLife>

    *   Search CPAN

        <http://search.cpan.org/dist/WWW-FMyLife/>

    *   FML (FMyLife)

        <http://www.fmylife.com>

SEE ALSO
    *   WWW::VieDeMerde

        Apparently supports more options right now. Mainly for French
        version but seems to support English as well.

LICENSE AND COPYRIGHT
    Copyright 2010 Sawyer X, Tamir Lousky.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for WWW-FMyLife
(format of DD/MM/YY)

0.15
        - patch allowing get_id and last of categories by Gavin Hanover!

0.14    29.01.10
        - disabled error tests

0.13    18.01.10
        - FML gave me a private key, i'm using fixed xml data i got using it

0.12    17.12.09
        - this time fixed tests number in 01-mixed_time.t for real

0.11    15.12.09
        - comment is now Int|EmptyComment, which is an empty hash
        - correct tests number in t/view/01-mixed_time.t
        - removed testing_done() in t/view/01-mixed_time.t

0.10    13.12.09
        - recompiled some tests, mixed them together
        - added more flags for skipping if item not defined
        - updated README (apparently not updated since 0.08)

0.09    05.12.09
        - fixed erred test to check for a general error
        - skip item if it's not defined in view/04-top_day.t

0.08    02.12.09
        - added "vote" attribute to Item
          (hopefully that will fix the failing tests)

0.07    13.09.09
        - fixed bug in test
        - fixed bug in top day/week/month
        - added flop and flop day/week/month + tests
        - removed hardcoded "no critic"s
        - fixed critic test to only run on RELEASE_TESTING
        - adding and fixing POD for Perl::Critic

0.06    07.09.09
        - added no critic for weird critic fails
        - changed AUTHORS section to AUTHOR (critic)
        - fixed spelling error

0.05    04.09.09
        - made immutable

0.04    03.09.09
        - adding top and top variations
         (top_day, top_week, top_month
        - adding random
        - adding tests for them
        - updating pod and fixing it up
        - refactored a lot
        - correcting repository in resources in META.yml

0.03    23.08.09
        - Some POD changes (also including link to WWW::VieDeMerde)

0.02    28.08.09
        - Fixing MANIFEST, META.yml files

0.01    28.08.09 w00t!
        - Created the basic skeleton
        - Added a lot of tests
        - Currently supports last()



Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name          => 'WWW::FMyLife',
    license              => 'perl',
    dist_author          => [
        q{Sawyer X <xsawyerx@cpan.org>},
        q{Tamir Lousky <tlousky@cpan.org>},
    ],
    dist_version_from    => 'lib/WWW/FMyLife.pm',
    build_requires   => {
        'Net::Ping'                 => 0,
        'Test::More'                => 0.88,
        'Test::Deep'                => 0,
        'Test::MockObject::Extends' => 0,
    },
    requires             => {
        'Moose'          => 0.86,
        'LWP::UserAgent' => 0,
        'XML::Simple'    => 2.18
    },
    recursive_test_files => 1,
    add_to_cleanup       => [ 'WWW-FMyLife-*' ],
    create_makefile_pl   => 'traditional',
    meta_merge           => {
        resources => {
            repository => 'http://github.com/xsawyerx/www-fmylife/tree/master',
        },
        keywords   => [ 'fml', 'fmylife' ],
    },
    create_readme        => 1,
    create_license       => 1,
);

$builder->create_build_script();


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.35_13
use ExtUtils::MakeMaker;
WriteMakefile
(
          'NAME' => 'WWW::FMyLife',
          'VERSION_FROM' => 'lib/WWW/FMyLife.pm',
          'PREREQ_PM' => {
                           'LWP::UserAgent' => '0',
                           'Moose' => '0.86',
                           'Net::Ping' => '0',
                           'Sub::Override' => '0',
                           'Test::Deep' => '0',
                           'Test::MockObject::Extends' => '0',
                           'Test::More' => '0.88',
                           'XML::Simple' => '2.18'
                         },
          'INSTALLDIRS' => 'site',
          'EXE_FILES' => [],
          'PL_FILES' => {},
          'test' => {
                      'TESTS' => 't/*.t t/data/*.t t/view/*.t'
                    }
        )
;


lib/WWW/FMyLife/Item.pm
package WWW::FMyLife::Item;

use Moose;
use Moose::Util::TypeConstraints;

our $VERSION = '0.05';

subtype 'FMLAuthor'
    => as 'HashRef'
    => where { exists shift->{'content'} };

subtype 'FMLVote'
    => as 'HashRef'
    => where {
        my $ref = shift;
        ( exists $ref->{'deserved'}     && exists $ref->{'agree'}     ) &&
        ( $ref->{'deserved'} =~ /^\d+$/ && $ref->{'agree'} =~ /^\d+$/ )
    };

subtype 'EmptyComment'
    => as 'HashRef'
    => where {
        my $ref = shift;
        scalar keys %{$ref} == 0;
    };

has 'id'            => ( is => 'rw', isa => 'Int'              );
has 'author'        => ( is => 'rw', isa => 'FMLAuthor'        );
has 'date'          => ( is => 'rw', isa => 'Str'              );
has 'category'      => ( is => 'rw', isa => 'Str'              );
has 'text'          => ( is => 'rw', isa => 'Str'              );
has 'vote'          => ( is => 'rw', isa => 'HashRef'          );
has 'deserved'      => ( is => 'rw', isa => 'Int'              );
has 'agree'         => ( is => 'rw', isa => 'Int'              );
has 'comments'      => ( is => 'rw', isa => 'Int|EmptyComment' );
has 'comments_flag' => ( is => 'rw', isa => 'Bool'             );

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 NAME

WWW::FMyLife::Item - Represents a single FMyLife.com Item

=head1 VERSION

Version 0.04

=head1 SYNOPSIS

    my $fml = WWW::FMyLife::Item->new( {
        author => 'me',
        text   => 'Stuff',
    } );

    ...

=head1 EXPORT

Nothing.

=head1 METHODS

=head2 new

Create a new item which can be submitted to FML.

=head1 AUTHORS

Sawyer X (XSAWYERX), C<< <xsawyerx at cpan.org> >>
Tamir Lousky (TLOUSKY), C<< <tlousky at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-www-fmylife at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-FMyLife>.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::FMyLife

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-FMyLife>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-FMyLife>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-FMyLife>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-FMyLife/>

=back

=head1 ACKNOWLEDGEMENTS

=head1 COPYRIGHT & LICENSE

Copyright 2009 Sawyer X, Tamir Lousky.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



lib/WWW/FMyLife/Vote.pm
package WWW::FMyLife::Vote;

use Moose;

our $VERSION = '0.02';

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 NAME

WWW::FMyLife::Vote - Represents an FMyLife.com vote

=head1 VERSION

Version 0.02

=head1 SYNOPSIS

    my $fml = WWW::FMyLife::Vote->new( {
        id   => '999',
        text => 'Stuff',
    } );

    ...

=head1 EXPORT

Nothing.

=head1 METHODS

=head2 new

Create a new vote relating to a post or other comment.

=head1 AUTHORS

Sawyer X (XSAWYERX), C<< <xsawyerx at cpan.org> >>
Tamir Lousky (TLOUSKY), C<< <tlousky at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-www-fmylife at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-FMyLife>.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::FMyLife

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-FMyLife>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-FMyLife>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-FMyLife>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-FMyLife/>

=back

=head1 ACKNOWLEDGEMENTS

=head1 COPYRIGHT & LICENSE

Copyright 2009 Sawyer X, Tamir Lousky.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



lib/WWW/FMyLife/Quote.pm
package WWW::FMyLife::Quote;

use Moose;

our $VERSION = '0.01';

1;

__END__

=head1 NAME

WWW::FMyLife::Comment - Represents a single FMyLife.com Quote

=head1 VERSION

Version 0.01

=head1 SYNOPSIS

    my $fml = WWW::FMyLife::Quote->new( {
        author => 'me',
        text   => 'Stuff',
    } );

    ...

=head1 EXPORT

Nothing.

=head1 METHODS

=head2 new

Create a new quote which can be submitted to FML.

=head1 AUTHORS

Sawyer X (XSAWYERX), C<< <xsawyerx at cpan.org> >>
Tamir Lousky (TLOUSKY), C<< <tlousky at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-www-fmylife at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-FMyLife>.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::FMyLife

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-FMyLife>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-FMyLife>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-FMyLife>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-FMyLife/>

=back

=head1 ACKNOWLEDGEMENTS

=head1 COPYRIGHT & LICENSE

Copyright 2009 Sawyer X, Tamir Lousky.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



lib/WWW/FMyLife/Comment.pm
package WWW::FMyLife::Comment;

use Moose;

our $VERSION = '0.02';

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 NAME

WWW::FMyLife::Comment - Represents an FMyLife.com comment

=head1 VERSION

Version 0.02

=head1 SYNOPSIS

    my $fml = WWW::FMyLife::Comment->new( {
        id   => '999',
        text => 'Stuff',
    } );

    ...

=head1 EXPORT

Nothing.

=head1 METHODS

=head2 new

Create a new comment relating to a post or other comment.

=head1 AUTHORS

Sawyer X (XSAWYERX), C<< <xsawyerx at cpan.org> >>
Tamir Lousky (TLOUSKY), C<< <tlousky at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-www-fmylife at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-FMyLife>.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::FMyLife

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-FMyLife>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-FMyLife>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-FMyLife>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-FMyLife/>

=back

=head1 ACKNOWLEDGEMENTS

=head1 COPYRIGHT & LICENSE

Copyright 2009 Sawyer X, Tamir Lousky.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



lib/WWW/FMyLife.pm
package WWW::FMyLife;

use Moose;
use XML::Simple;
use LWP::UserAgent;
use WWW::FMyLife::Item;

our $VERSION = '0.14';

has 'username' => ( is => 'rw', isa => 'Str' );
has 'password' => ( is => 'rw', isa => 'Str' );

has 'language' => ( is => 'rw', isa => 'Str', default => 'en'       );
has 'token'    => ( is => 'rw', isa => 'Str', default => q{}        );
has 'key'      => ( is => 'rw', isa => 'Str', default => 'readonly' );

# XXX: is there a point to this? is this wannabe caching?
has 'pages'    => ( is => 'rw', isa => 'Int' );

has 'api_url'  => (
    is      => 'rw',
    isa     => 'Str',
    default => 'http://api.betacie.com',
);

has 'module_error' => (
    is      => 'rw',
    isa     => 'Str',
    clearer => 'clear_module_error',
);

has 'fml_errors'   => (
    is => 'rw',
    isa => 'ArrayRef[Str]',
    clearer => 'clear_fml_errors',
);

has 'error'        => (
    is      => 'rw',
    isa     => 'Bool',
    default => 0,
);

has 'agent'    => (
    is      => 'rw',
    isa     => 'Object',
    default => sub { LWP::UserAgent->new(); },
);

# Credentials sub: sets username and password as an array
sub credentials {
    my ( $self, $user, $pass ) = @_;
    $self->username ( $user );
    $self->password ( $pass );

    return;
}

sub top {
    my ( $self, $opts ) = @_;
    my @items = $self->_parse_options( $opts, 'top' );
    return @items;
}

sub top_day {
    my ( $self, $opts ) = @_;
    my @items = $self->_parse_options( $opts, 'top_day' );
    return @items;
}

sub top_week {
    my ( $self, $opts ) = @_;
    my @items = $self->_parse_options( $opts, 'top_week' );
    return @items;
}

sub top_month {
    my ( $self, $opts ) = @_;
    my @items = $self->_parse_options( $opts, 'top_month' );
    return @items;
}

sub flop {
    my ( $self, $opts ) = @_;
    my @items = $self->_parse_options( $opts, 'flop' );
    return @items;
}

sub flop_day {
    my ( $self, $opts ) = @_;
    my @items = $self->_parse_options( $opts, 'flop_day' );
    return @items;
}

sub flop_week {
    my ( $self, $opts ) = @_;
    my @items = $self->_parse_options( $opts, 'flop_week' );
    return @items;
}

sub flop_month {
    my ( $self, $opts ) = @_;
    my @items = $self->_parse_options( $opts, 'flop_month' );
    return @items;
}

sub last {
    my ( $self, $opts ) = @_;
    my $type = 'last';

    if ( ref $opts eq 'HASH' && $opts->{'category'} ) {
        $type = $opts->{'category'};
    }

    my @items = $self->_parse_options( $opts, $type );
    return @items;
}

sub get_id {
    my ( $self, $id, $opts ) = @_;
    $opts->{'page'} = '/nocomment';
    my @items = $self->_parse_options( $opts, $id  );
    return @items;
}

sub random {
    my $self = shift;
    my $xml  = $self->_fetch_data('/view/random');
    my $item = $self->_parse_item_as_object($xml);
    return $item;
}

sub _parse_options {
    my ( $self, $opts, $add_url ) = @_;
    my ( $as,   $page );

    if ( ref $opts eq 'HASH' ) {
        $as   = $opts->{'as'};
        $page = $opts->{'page'};
    } else {
        $page = $opts;
    }

    $as   ||= 'object';
    $page ||= q{};

    my %types = (
        object => sub { return $self->_parse_items_as_object(@_) },
        text   => sub { return $self->_parse_items_as_text  (@_) },
        data   => sub { return $self->_parse_items_as_data  (@_) },
    );

    my $xml = $self->_fetch_data("/view/$add_url/$page");

    $xml || return;

    if ( my $id = $xml->{'items'}{'item'}{'id'} ) {
        $xml->{'items'}{'item'} = { $id => $xml{'items'}{'item'} };
        $xml->{'pages'}         = 1;
    }

    $self->pages( $xml->{'pages'} );

    my @items = $types{$as}->($xml);

    return @items;
}

sub _fetch_data {
    my ( $self, $add_to_url ) = @_;

    my $res = $self->agent->post(
        $self->api_url . $add_to_url, {
            key      => $self->key,
            language => $self->language,
        },
    );

    $self->error(0);
    $self->clear_fml_errors;
    $self->clear_module_error;

    if ( ! $res->is_success ) {
        $self->error(1);
        $self->module_error( $res->status_line );
        return;
    }

    my $xml = XMLin( $res->decoded_content );

    if ( my $raw_errors = $xml->{'errors'}->{'error'} ) {
        my $array_errors =
            ref $raw_errors eq 'ARRAY' ? $raw_errors : [ $raw_errors ];

        $self->error(1);
        $self->fml_errors($array_errors);
        return;
    }

    return $xml;
}

sub _parse_item_as_object {
    # this parses a single item
    my ( $self, $xml ) = @_;

    my %item_data = %{ $xml->{'items'}{'item'} };
    my $item      = WWW::FMyLife::Item->new();

    foreach my $attr ( keys %item_data ) {
        $item->$attr( $item_data{$attr} );
    }

    return $item;
}

sub _parse_items_as_object {
    # this parses multiple items
    my ( $self, $xml ) = @_;
    my @items;

    while ( my ( $id, $item_data ) = each %{ $xml->{'items'}{'item'} } ) {
        my $item = WWW::FMyLife::Item->new(
            id => $id,
        );

        foreach my $attr ( keys %{$item_data} ) {
            $item->$attr( $item_data->{$attr} );
        }

        push @items, $item;
    }

    return @items;
}

sub _parse_items_as_text {
    my ( $self, $xml ) = @_;
    my @items = map { $_->{'text'} } values %{ $xml->{'items'}{'item'} };
    return @items;
}

sub _parse_items_as_data {
    my ( $self, $xml ) = @_;
    my $itemsref       = $xml->{'items'}{'item'};
    my @items          = map +{ $_ => $itemsref->{$_} }, keys %{$itemsref};
    return @items;
}

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 NAME

WWW::FMyLife - Obtain FMyLife.com anecdotes via API

=head1 VERSION

Version 0.14

=head1 SYNOPSIS

THIS MODULE IS STILL UNDER INITIAL DEVELOPMENT! BE WARNED!

    use WWW::FMyLife;

    my $fml = WWW::FMyLife->new();
    print map { "Items: $_\n" } $fml->last( { as => text' } );

=head1 DESCRIPTION

This module fetches FMyLife.com (FML) anecdotes, comments, votes and more via API, comfortably and in an extensible manner.

    my @items = $fml->top_daily();
    foreach my $item (@items) {
        my $item_id      = $item->id;
        my $item_content = $item->content;
        print "[$item_id] $item_content\n";
    }

    print $fml->random()->text, "\n";
    ...

=head1 EXPORT

This module exports nothing.

=head1 SUBROUTINES/METHODS

=head2 last()

Fetches the last quotes. Can accept a hashref that indicates the formatting:

    # returns an array of WWW::FMyLife::Item objects
    $fml->last();

    # or more explicitly
    $fml->last( { as => 'object' } ); # same as above
    $fml->last( { as => 'text'   } ); # returns an array of text anecdotes
    $fml->last( { as => 'data'   } ); # returns an array of hashes of anecdotes

You can also specify which page you want:

    # return 1st page
    my @last = fml->last();

    # return 5th page
    my @last = $fml->last(5);

    # same
    my @last = $fml->last( { page => 5 } );

And options can be mixed:

    my @not_so_last = $fml->last( { as => 'text', page => 50 } );

=head2 random

This method gets a single random quote as an object.

=head2 top

This method works the same as the last() method, only it fetches the top quotes.

This method, as for its variations, can format as an object, text or data.

=head2 top_day

This method works the same as the last() method, only it fetches the top quotes.

This specific variant fetches the top anecdotes from the last day.

=head2 top_week

This method works the same as the last() method, only it fetches the top quotes.

This specific variant fetches the top anecdotes from the last week.

=head2 top_month

This method works the same as the last() method, only it fetches the top quotes.

This specific variant fetches the top anecdotes from the last month.

=head2 flop

Fetches the flop quotes.

This method, as for its variations, can format as an object, text or data.

=head2 flop_day

Fetches the flop quotes of the day.

=head2 flop_week

Fetches the flop quotes of the week.

=head2 flop_month

Fetches the flop quotes of the month.

=head2 credentials( $username, $password ) (NOT YET FULLY IMPLEMENTED)

WARNING: THIS HAS NOT YET BEEN IMPLEMENTED.

THE TESTS HAVE BEEN DISABLED FOR NOW, PLEASE WAIT FOR A MORE ADVANCED VERSION.

Sets credentials for members.

    $fml->credentials( 'foo', 'bar' );

    # same thing
    $fml->username('foo');
    $fml->password('bar');

=head1 AUTHOR

Sawyer X (XSAWYERX), C<< <xsawyerx at cpan.org> >>

Tamir Lousky (TLOUSKY), C<< <tlousky at cpan.org> >>

=head1 DEPENDENCIES

L<Moose>

L<XML::Simple>

L<LWP::UserAgent>

=head1 BUGS AND LIMITATIONS

Please report any bugs or feature requests to C<bug-www-fmylife at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-FMyLife>.

You can also use the Issues Tracker on Github @ L<http://github.com/xsawyerx/www-fmylife/issues>.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::FMyLife

You can also look for information at:

=over 4

=item * Our Github!

L<http://github.com/xsawyerx/www-fmylife/tree/master>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-FMyLife>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-FMyLife>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-FMyLife>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-FMyLife/>

=item * FML (FMyLife)

L<http://www.fmylife.com>

=back

=head1 SEE ALSO

=over 4

=item * L<WWW::VieDeMerde>

Apparently supports more options right now. Mainly for French version but seems to support English as well.

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2010 Sawyer X, Tamir Lousky.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

